openjdk/mmtkBarrierSetAssembler_x86.cpp

```diff
@@ -114,33 +114,35 @@ void MMTkBarrierSetAssembler::eden_allocate(MacroAssembler* masm, Register threa
 #ifdef MMTK_ENABLE_GLOBAL_ALLOC_BIT
     Register tmp3 = rscratch2;
     Register tmp2 = rscratch1;
-    Register tmp4 = t1;
-    assert_different_registers(obj, tmp2, tmp3, tmp4, rcx);
+    assert_different_registers(obj, tmp2, tmp3, rcx);
+
+    __ push(tmp2);
```

Comment:
Yes, we can destroy the values in rscratch1/2 if the caller saves their values (the caller should save them if it wants to use them after calling another function). However, in this case I find that I have to save rscratch2, otherwise something will be broken.

Anyway, I now expicitly use rdi here instead of rscratch2.